src/usng.ts

```diff
@@ -0,0 +1,1396 @@
+    LLtoUTMwithNS: function (lat,lon,utmcoords,zone) {
+        this.LLtoUTM(lat,lon,utmcoords,zone);
+        if (utmcoords[1] < 0) {
+            utmcoords[1] += this.NORTHING_OFFSET;
+            utmcoords[3] = 'S';
+        } else {
+            utmcoords[3] = 'N';
+        }
+    },
+
+    /***************** convert latitude, longitude to UTM  *******************
+    Uses N or S indicator instead of returning a negative Northing value
+    ***************************************************************************/
+    UTMtoLLwithNS: function (UTMNorthing, UTMEasting, UTMZoneNumber, accuracy, NorSIndicator) {
+        var result;
+        if (NorSIndicator === 'S') {
+            result = this.UTMtoLL(UTMNorthing - this.NORTHING_OFFSET, UTMEasting, UTMZoneNumber, accuracy);
+        } else {
+            result = this.UTMtoLL(UTMNorthing, UTMEasting, UTMZoneNumber, accuracy);
+        }
+
+        return result;
+    },
+
+    serializeUPS({northPole, easting, northing}) {
+        const zoneLetter = northPole ? (easting < 2000000 ? 'Y' : 'Z')
+            : (easting < 2000000 ? 'A' : 'B')
+        return `${zoneLetter} ${easting}mE ${northing}mN`
+    },
+
+    deserializeUPS(str) {
+        const regExp = /^\s*([ABYZ])\s+(\d+)mE\s+(\d+)mN\s*$/i
+        try {
+            const [, zoneLetter, easting, northing] = regExp.exec(str)
+            this.UPStoLL({
+                northPole: true,
+                easting: Number(easting),
+                northing: Number(northing)
+            })
+            return {
+                northPole: includes(["Y", "Z"], zoneLetter.toUpperCase()),
+                easting: Number(easting),
+                northing: Number(northing)
+            }
+        } catch(err) {
+            throw new Error( `Invalid UPS String: ${str}`)
+        }
+    },
+
+    /***************** convert latitude, longitude to UPS  *******************
+     Uses "northPole" boolean instead of returning a negative Northing value
+
+     Input latitude is expected to be [-90, 90].
+     Input longitude is expected to be [-180, 180].
+
+     This returns an object describing the UPS coordinates:
+     { northing: number, easting: number, northPole: bool }
+
+     On invalid input throws an error.
+    ***************************************************************************/
+    LLtoUPS: function(lat, lon) {
+        const isLatLonCoordinatesPairInvalid = () => {
+            return !(typeof lat === "number"
+                && typeof lon === "number"
+                && lat >= -90 && lat <= 90
+                && lon >= -180 && lon <= 180)
+        }
+        if (isLatLonCoordinatesPairInvalid()) {
+            throw new Error('Invalid Lat/Lon coordinates: ' + lat + ', ' + lon)
+        }
+        const northPole = lat > 0
+        const tau = Math.tan(Math.abs(lat) * this.DEG_2_RAD)
+        const taup = taupf.call(this, tau)
+        const rhoStep1 = hypot(1, taup) + Math.abs(taup)
+        let rhoStep2
```

Comment:
:question: Are the 10 lines starting from 542 reallf more readable than the following 5 lines:
```javascript
 const rhoStep2 = taup >= 0
            ? (Math.abs(lat) !== 90
                ? 1/rhoStep1
                : 0)
            : rhoStep1
```

Reply:
I would argue that the ternary is better, especially since it lets us keep a variable immutable instead of mutable, but that view conflicts with [this comment](https://github.com/codice/usng.js/pull/18#discussion_r220310828). Then again, the aforementioned comment also suggests simply factoring the branching calculation out into a separate function might be an option to satisfy these concerns and those in the comment. @coyotesqrl, @dimitry-sherman, thoughts?

Reply:
Commented on the original complaint here: https://github.com/codice/usng.js/pull/18#discussion_r220703263